modules/core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/model/BaseSession.java

```diff
@@ -0,0 +1,173 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Florent Guillaume
+ */
+package org.nuxeo.ecm.core.model;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.nuxeo.ecm.core.api.security.ACE;
+import org.nuxeo.ecm.core.api.security.ACL;
+import org.nuxeo.ecm.core.api.security.ACP;
+import org.nuxeo.ecm.core.api.security.Access;
+import org.nuxeo.ecm.core.api.security.SecurityConstants;
+import org.nuxeo.ecm.core.api.security.impl.ACPImpl;
+import org.nuxeo.ecm.core.blob.DocumentBlobManager;
+import org.nuxeo.ecm.core.query.QueryFilter;
+import org.nuxeo.runtime.api.Framework;
+
+/**
+ * Common code for VCS and DBS {@link Session} implementations.
+ *
+ * @since 11.1
+ */
+public abstract class BaseSession implements Session<QueryFilter> {
+
+    protected final Repository repository;
+
+    protected BaseSession(Repository repository) {
+        this.repository = repository;
+    }
+
+    protected DocumentBlobManager getDocumentBlobManager() {
+        return Framework.getService(DocumentBlobManager.class);
+    }
+
+    protected void notifyAfterCopy(Document doc) {
+        getDocumentBlobManager().notifyAfterCopy(doc);
+    }
+
+    /*
+     * ----- Common ACP code -----
+     */
+
+    protected void checkNegativeAcl(ACP acp) {
+        if (acp == null || isNegativeAclAllowed()) {
+            return;
+        }
+        for (ACL acl : acp.getACLs()) {
+            if (acl.getName().equals(ACL.INHERITED_ACL)) {
+                continue;
+            }
+            for (ACE ace : acl.getACEs()) {
+                if (ace.isGranted()) {
+                    continue;
+                }
+                String permission = ace.getPermission();
+                if (permission.equals(SecurityConstants.EVERYTHING)
+                        && ace.getUsername().equals(SecurityConstants.EVERYONE)) {
+                    continue;
+                }
+                // allow Write, as we're sure it doesn't include Read/Browse
+                if (permission.equals(SecurityConstants.WRITE)) {
+                    continue;
+                }
+                throw new IllegalArgumentException("Negative ACL not allowed: " + ace);
+            }
+        }
+    }
+
+    /**
+     * Gets the ACP for the document (without any inheritance).
+     *
+     * @param doc the document
+     * @return the ACP
+     */
+    public abstract ACP getACP(Document doc);
```

Comment:
it seems it was protected when it was in the implementations, is it worth a @since 11.1 ?